chall04/gbudau.py

```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/env python3
+
+import sys
+
+STDIN = 0
+FILE = 1
+
+def read_error(filename):
+    print("{}: {}: Can't read file".format(sys.argv[0], filename))
+
+def space_error(filename):
+    print("{}: {}: Not enough space in the given shelves".format(sys.argv[0], filename))
+
+def fill_bookshelves(filename, bookshelves, books):
+    # Convert books and bookshelves to integers
+    # Sort from max to min
+    books = [int(i) for i in books]
+    bookshelves = [int(i) for i in bookshelves]
+    books.sort(reverse = True)
+    bookshelves.sort(reverse = True)
+
+    used_bookshelves = []
+    while books and bookshelves and ((bookshelves[0] >= books[0]) or (used_bookshelves and used_bookshelves[0] >= books[0])):
+        if  used_bookshelves and used_bookshelves[0] >= books[0]:
+            used_bookshelves[0] -= books[0]
+            books.pop(0)
+            used_bookshelves.sort(reverse = True)
+        else:
+            used_bookshelves.append(bookshelves[0] - books[0])
+            used_bookshelves.sort(reverse = True)
+            books.pop(0)
+            bookshelves.pop(0)
```

Comment:
Really interresting approach, thank you for sharing this.

I tried one case with the example that might be wrong, it returns 2 instead of 3, i edited the function a bit to accomodate to my script so maybe i got it wrong.
```
100 1000 2000
1500 book1
500 book2
1000 book3
100 book4
```

```
def fill_bookshelves(filename, shelves, books):
    used_shelves = 1
    books = sorted([int(i) for i in books], reverse=True)
    shelves = sorted([int(i) for i in shelves], reverse=True)
    shelves = iter(shelves)
    shelf = next(shelves)
    for book in books:
        if shelf >= book:
            shelf -= book
        else:
            try:
                shelf = next(shelves)
                used_shelves += 1
                if shelf < book:
                    space_error(filename)
                    break
            except StopIteration:
                break
    else:
        print(used_shelves)
```